module/src/hooks/useInfinitePaging.ts

```diff
@@ -0,0 +1,214 @@
+  hasFinished: false,
+  error: undefined,
+});
+
+interface IFetchAction<T> {
+  type: 'fetch';
+  newItems: T[];
+  hasFinished: boolean;
+  nextPageToken?: PageToken;
+  getKey: InfinitePagingKey<T>;
+}
+interface IErrorAction {
+  type: 'error';
+  error: any;
+}
+
+interface IResetAction<T> {
+  type: 'reset';
+  initialItems: T[];
+  getKey: InfinitePagingKey<T>;
+}
+
+interface IInsertAction<T> {
+  type: 'insert';
+  newItems: T[];
+  getKey: InfinitePagingKey<T>;
+}
+
+interface IRemoveAction {
+  type: 'remove';
+  key: string | number;
+}
+
+type InfinitePagingAction<T> = IFetchAction<T> | IErrorAction | IResetAction<T> | IInsertAction<T> | IRemoveAction;
+
+function useInfinitePagingReducer<T>(state: IUseInfinitePagingState<T>, action: InfinitePagingAction<T>): IUseInfinitePagingState<T> {
+  switch (action.type) {
+    case 'fetch': {
+      const items = {
+        ...state.items,
+        ...Arrays.arrayToDictionary(action.newItems, action.getKey),
+      };
+
+      return {
+        items,
+        nextPageToken: action.nextPageToken,
+        error: undefined,
+        hasFinished: action.hasFinished,
+      };
+    }
+    case 'reset': {
+      return initialState(Arrays.arrayToDictionary(action.initialItems, action.getKey));
+    }
+    case 'error': {
+      return {
+        ...state,
+        error: action.error,
+      };
+    }
+    case 'insert': {
+      const items = {
+        ...state.items,
+        ...Arrays.arrayToDictionary(action.newItems, action.getKey),
+      };
+      return {
+        ...state,
+        items,
+      };
+    }
+    case 'remove': {
+      const items = {
+        ...state.items,
+      };
+
+      delete items[action.key];
+
+      return {
+        ...state,
+        items,
+      };
+    }
+    default: {
+      return state;
+    }
+  }
+}
+
+/** A hook used to create an infinite paging state - it takes a fetch callback that returns some data that is added to a piece of state that's returned as on big array */
+export function useInfinitePaging<T>(
+  fetch: (pageToken?: PageToken) => Promise<IInfinitePagingResult<T>>,
+  getKey: InfinitePagingKey<T>,
+  settings: IUseInfinitePagingSettings<T> = {}
+) {
+  const [state, dispatch] = React.useReducer<React.Reducer<IUseInfinitePagingState<T>, InfinitePagingAction<T>>>(
+    useInfinitePagingReducer,
+    initialState(Arrays.arrayToDictionary<T>(settings.initialItems || [], getKey))
+  );
+
+  const [isFetching, setIsFetching] = React.useState(false);
+
+  const fetcher = React.useCallback(
+    async (fetchPageToken?: PageToken) => {
+      setIsFetching(true);
+
+      try {
+        const response = await fetch(fetchPageToken);
+
+        if (!response) {
+          throw new Error('No response returned from fetch in useInfinitePaging');
+        }
+
+        const noReturnedItems = !response.data || response.data.length === 0;
+        const responseSmallerThanPageSize = settings.pageSize && response.data.length < settings.pageSize;
```

Comment:
The above line suggests that `response.data` might be falsy, so `response.data.length` seems risky here.

Suggestion:
```
const responseSmallerThanPageSize = !noReturnedItems && settings.pageSize && response.data.length < settings.pageSize;

```